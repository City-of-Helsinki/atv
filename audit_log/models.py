from typing import Optional

from django.db import models
from django.utils import timezone
from django.utils.translation import gettext_lazy as _


class AuditLogEntry(models.Model):
    is_sent = models.BooleanField(default=False, verbose_name=_("is sent"))
    message = models.JSONField(verbose_name=_("message"))
    created_at = models.DateTimeField(auto_now_add=True, verbose_name=_("created at"))

    class Meta:
        verbose_name_plural = "audit log entries"

    def __str__(self):
        return ' '.join(
            [
                _safe_get(self.message, "audit_event", "date_time"),
                _safe_get(self.message, "audit_event", "actor", "role"),
                _safe_get(self.message, "audit_event", "actor", "user_id"),
                _safe_get(self.message, "audit_event", "operation"),
                _safe_get(self.message, "audit_event", "target", "type").upper(),
                _safe_get(self.message, "audit_event", "target", "id"),
                "(SENT)" if self.is_sent else "(NOT SENT)",
            ]
        )

    @property
    def timestamp(self):
        return (
            _safe_get(self.message, "audit_event", "date_time", empty_none=True)
            or timezone.now()
        )


def _safe_get(value: dict, *keys: str, empty_none=False) -> Optional[str]:
    """Look up a nested key in the given dict, or return "UNKNOWN" on KeyError."""
    for key in keys:
        try:
            value = value[key]
        except KeyError:
            return "UNKNOWN" if not empty_none else None
    return str(value)
