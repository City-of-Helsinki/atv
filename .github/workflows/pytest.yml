name: Python tests

on:
  push:
    branches: [main, release]
  pull_request:

env:
  SECRET_KEY: topsecret123
  FIELD_ENCRYPTION_KEYS: 4b52331c68c8b32dcfba241d6d6b23466b80a8b6d0cea91ea64eb1e23c44dff4

jobs:
  pytest:
    name: Python tests
    runs-on: ubuntu-latest

    services:
      postgres:
        image: postgres:11
        ports:
          - 5432:5432
        options: >-
          --health-cmd pg_isready
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5
        env:
          POSTGRES_USER: atv
          POSTGRES_PASSWORD: atv
          POSTGRES_DB: atv

    steps:
      - name: Check out repository
        uses: actions/checkout@v2

      - name: Set up Python 3.9
        uses: actions/setup-python@v2
        with:
          python-version: '3.9'

      - name: Cache pip packages
        uses: actions/cache@v2
        env:
          cache-name: cache-pip-modules
        with:
          path: ~/.pip-cache
          key: ${{ runner.os }}-build-${{ env.cache-name }}-${{ hashFiles('**/requirements.txt') }}-${{ hashFiles('**/requirements-dev.txt') }}
          restore-keys: |
            ${{ runner.os }}-build-${{ env.cache-name }}-
            ${{ runner.os }}-build-
            ${{ runner.os }}-

      - name: Install dependencies
        run: pip install -r requirements.txt -r requirements-dev.txt codecov

      - name: Run tests
        run: pytest -ra -vv --doctest-modules --cov=.
        env:
          DATABASE_URL: postgres://atv:atv@localhost:5432/atv

      - name: Coverage
        run: codecov
