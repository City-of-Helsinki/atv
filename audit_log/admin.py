from django.contrib import admin

from audit_log.models import AuditLogEntry


@admin.register(AuditLogEntry)
class AuditLogEntryAdmin(admin.ModelAdmin):
    fields = ("id", "created_at", "message")
    list_display = ("__str__", "created_at")
    list_filter = ("created_at", "is_sent")
    readonly_fields = ("id", "created_at", "message")

    def has_delete_permission(self, request, obj=None):
        return False

    def has_add_permission(self, request):
        return False

    def has_change_permission(self, request, obj=None):
        return False
