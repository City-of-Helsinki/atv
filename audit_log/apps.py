from django.apps import AppConfig
from django.utils.translation import gettext_lazy as _


class AuditLogConfig(AppConfig):
    default_auto_field = "django.db.models.BigAutoField"
    name = "audit_log"
    verbose_name = _("audit log")
