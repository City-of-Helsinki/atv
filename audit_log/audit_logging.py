from datetime import datetime, timezone
from typing import Callable, Optional, Union

from django.conf import settings
from django.contrib.auth import get_user_model
from django.contrib.auth.models import AnonymousUser
from django.db.models import Model
from django.db.models.base import ModelBase

from audit_log.enums import Operation, Role, Status
from audit_log.models import AuditLogEntry

User = get_user_model()


def _now() -> datetime:
    """Returns the current time in UTC timezone."""
    return datetime.now(tz=timezone.utc)


def _iso8601_date(time: datetime) -> str:
    """Formats the timestamp in ISO-8601 format, e.g. '2020-06-01T00:00:00.000Z'."""
    return f"{time.replace(tzinfo=None).isoformat(timespec='milliseconds')}Z"


def log(
    actor: Optional[Union[User, AnonymousUser]],
    actor_backend: str,
    operation: Operation,
    target: Union[Model, ModelBase],
    status: Status = Status.SUCCESS,
    get_time: Callable[[], datetime] = _now,
    ip_address: str = "",
    additional_information: str = "",
    service=None,
    view_name: str = "",
    lookup_field: str = "",
):
    """
    Write an event to the audit log.

    Each audit log event has an actor (or None for system events),
    an operation(e.g. READ or UPDATE), the target of the operation
    (a Django model instance), status (e.g. SUCCESS), and a timestamp.
    """
    current_time = get_time()
    user_id_field_name = getattr(actor, "audit_log_id_field", "pk")
    user_id = str(getattr(actor, user_id_field_name, None) or "")

    if actor is None:
        role = Role.SYSTEM
    elif actor.is_superuser:
        role = Role.ADMIN
    elif isinstance(actor, AnonymousUser):
        role = Role.ANONYMOUS
    elif actor.id == target.pk:
        role = Role.OWNER
    else:
        role = Role.USER

    actor_data = {
        "role": role.value,
        "user_id": user_id,
        "provider": actor_backend if actor_backend else "",
        "ip_address": ip_address,
    }
    if service:
        actor_data["service"] = service.name
        actor_data["authentication"] = (
            "API-Key" if hasattr(actor, "service_api_key") else "Token"
        )
    target_id = _get_target_id(target)
    message = {
        "audit_event": {
            "origin": settings.AUDIT_LOG_ORIGIN,
            "status": status.value,
            "date_time_epoch": int(current_time.timestamp() * 1000),
            "date_time": _iso8601_date(current_time),
            "actor": actor_data,
            "operation": operation.value,
            "additional_information": additional_information,
            "target": {
                "type": _get_target_type(target),
                "id": target_id,
                "lookup_field": lookup_field if target_id else "",
                "endpoint": view_name,
            },
        },
    }
    AuditLogEntry.objects.create(message=message)


def _get_target_id(target: Union[Model, ModelBase]) -> str:
    if isinstance(target, ModelBase):
        return ""
    field_name = getattr(target, "audit_log_id_field", "pk")
    audit_log_id = getattr(target, field_name, "")
    return str(audit_log_id)


def _get_target_type(target: Union[Model, ModelBase]) -> str:
    return (
        str(target.__class__.__name__)
        if isinstance(target, Model)
        else str(target.__name__)
    )
