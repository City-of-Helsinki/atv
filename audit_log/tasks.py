import logging
from datetime import timedelta

from django.conf import settings
from django.utils import timezone
from elasticsearch import Elasticsearch

from audit_log.models import AuditLogEntry

ES_GOOD_STATUSES = ["created", "updated"]


logger = logging.getLogger(__name__)


def send_audit_log_entries():
    """Send audit logs to Elasticsearch and mark them as sent."""
    if not settings.ENABLE_SEND_AUDIT_LOG:
        logger.info("Sending audit logs not enabled, skipping.")
        return

    if not (settings.ELASTIC_HOST and settings.ELASTIC_AUDIT_LOG_INDEX):
        logger.warning(
            "Trying to send audit logs to Elasticsearch without proper configuration,"
            " process skipped"
        )
        return

    es = Elasticsearch(
        host=settings.ELASTIC_HOST,
        port=settings.ELASTIC_PORT,
        use_ssl=settings.ELASTIC_SSL,
        http_auth=(settings.ELASTIC_USERNAME, settings.ELASTIC_PASSWORD),
    )
    entries = AuditLogEntry.objects.filter(is_sent=False).order_by("created_at")

    if settings.ELASTIC_CREATE_DATA_STREAM:
        es.indices.create_data_stream(name=settings.ELASTIC_AUDIT_LOG_INDEX, ignore=400)

    for entry in entries.iterator():
        # Elastic data stream require @timestamp
        body = {"@timestamp": entry.timestamp, **entry.message}

        response = es.index(
            index=settings.ELASTIC_AUDIT_LOG_INDEX,
            body=body,
        )
        if response.get("result") in ES_GOOD_STATUSES:
            AuditLogEntry.objects.filter(pk=entry.pk).update(is_sent=True)

    if count := entries.count():
        logger.info(f"Sent {count} audit logs to elastic search.")


def clear_audit_log_entries(days_to_keep=30):
    """Remove sent audit log entries, which are older than `days_to_keep` days."""
    if not settings.CLEAR_AUDIT_LOG_ENTRIES:
        logger.info("Clearing audit logs not enabled, skipping.")
        return

    sent_entries = AuditLogEntry.objects.filter(
        is_sent=True, created_at__lte=(timezone.now() - timedelta(days=days_to_keep))
    )
    if count := sent_entries.count():
        sent_entries.delete()
        logger.info(
            f"Cleared {count} sent audit logs, which were older than {days_to_keep}"
            " days."
        )
