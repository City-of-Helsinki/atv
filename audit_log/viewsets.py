from contextlib import contextmanager
from copy import copy
from typing import Optional, Union

from django.conf import settings
from django.contrib.auth import get_user_model
from django.contrib.auth.models import AnonymousUser
from django.db import transaction
from django.db.models import Model
from django.db.models.base import ModelBase
from rest_framework.exceptions import NotAuthenticated, PermissionDenied
from rest_framework.viewsets import ModelViewSet

from audit_log import audit_logging
from audit_log.enums import Operation, Status

User = get_user_model()


class AuditLoggingModelViewSet(ModelViewSet):
    method_to_operation = {
        "POST": Operation.CREATE,
        "GET": Operation.READ,
        "PUT": Operation.UPDATE,
        "PATCH": Operation.UPDATE,
        "DELETE": Operation.DELETE,
    }
    created_instance: Optional[Model] = None

    def permission_denied(self, request, message=None, code=None):
        audit_logging.log(
            self._get_actor(),
            self._get_actor_backend(),
            self._get_operation(),
            self._get_target(),
            Status.FORBIDDEN,
            ip_address=self._get_ip_address(),
        )
        super().permission_denied(request, message, code)

    def retrieve(self, request, *args, **kwargs):
        with self.record_action():
            return super().retrieve(request, *args, **kwargs)

    @transaction.atomic
    def perform_create(self, serializer):
        with self.record_action():
            super().perform_create(serializer)
            self.created_instance = serializer.instance

    @transaction.atomic
    def perform_update(self, serializer):
        with self.record_action():
            super().perform_update(serializer)

    @transaction.atomic
    def perform_destroy(self, instance):
        target = copy(instance)  # Will be destroyed, so we must save it
        with self.record_action(target=target):
            super().perform_destroy(instance)

    @contextmanager
    def record_action(self, target: Optional[Model] = None):
        """
        This context manager will run the managed code in a transaction and writes
        a new audit log entry in the same transaction. If an exception is raised,
        the transaction will be rolled back. If the user has no permission to perform
        the given action, a "FORBIDDEN" audit log event will be recorded.
        """
        actor = copy(self._get_actor())  # May be destroyed if actor is also the target
        actor_backend = self._get_actor_backend()
        operation = self._get_operation()
        try:
            with transaction.atomic():
                yield
                audit_logging.log(
                    actor,
                    actor_backend,
                    operation,
                    target or self._get_target(),
                    ip_address=self._get_ip_address(),
                )
        except (NotAuthenticated, PermissionDenied):
            audit_logging.log(
                actor,
                actor_backend,
                operation,
                target or self._get_target(),
                Status.FORBIDDEN,
                ip_address=self._get_ip_address(),
            )
            raise

    def _get_actor(self) -> Union[User, AnonymousUser]:
        return self.request.user

    def _get_actor_backend(self) -> str:
        return self.request.session.get("_auth_user_backend", "")

    def _get_operation(self) -> Operation:
        return self.method_to_operation.get(self.request.method, Operation.READ)

    def _get_target(self) -> Optional[Union[Model, ModelBase]]:
        target = None
        lookup_value = self.kwargs.get(self.lookup_field, None)
        if lookup_value is not None:
            target = self.queryset.model.objects.filter(
                **{self.lookup_field: lookup_value}
            ).first()
        return target or self.created_instance or self.queryset.model()

    def _get_ip_address(self) -> str:
        client_ip = None

        if settings.USE_X_FORWARDED_FOR:
            forwarded_for = self.request.META.get("HTTP_X_FORWARDED_FOR", "")
            client_ip = forwarded_for.split(",")[0] or None

        if not client_ip:
            client_ip = self.request.META.get("REMOTE_ADDR")

        return client_ip
