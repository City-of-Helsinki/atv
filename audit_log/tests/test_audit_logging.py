from datetime import datetime, timezone

import pytest
from django.contrib.auth.models import AnonymousUser
from freezegun import freeze_time

from audit_log import audit_logging
from audit_log.enums import Operation, Status
from audit_log.models import AuditLogEntry


def test_log_origin(user):
    audit_logging.log(user, "", Operation.READ, user)

    message = AuditLogEntry.objects.first().message
    assert message["audit_event"]["origin"] == "atv"


@freeze_time("2020-06-01T00:00:00Z")
def test_log_current_timestamp(user):
    logging_datetime = datetime.now(tz=timezone.utc)
    audit_logging.log(user, "", Operation.READ, user)

    message = AuditLogEntry.objects.first().message
    logged_date_from_date_time_epoch = datetime.fromtimestamp(
        int(message["audit_event"]["date_time_epoch"]) / 1000, tz=timezone.utc
    )
    assert logging_datetime == logged_date_from_date_time_epoch
    logged_date_from_date_time = datetime.strptime(
        message["audit_event"]["date_time"], "%Y-%m-%dT%H:%M:%S.%f%z"
    )
    assert logging_datetime == logged_date_from_date_time


def test_log_actor_uuid(fixed_datetime, user, user_factory, snapshot):
    other_user = user_factory()
    audit_logging.log(
        user,
        "",
        Operation.READ,
        other_user,
        get_time=fixed_datetime,
        ip_address="192.168.1.1",
    )

    message = AuditLogEntry.objects.first().message
    assert message["audit_event"]["actor"]["user_id"] == str(user.uuid)
    snapshot.assert_match(message)


@pytest.mark.parametrize("operation", list(Operation))
def test_log_owner_operation(fixed_datetime, user, operation, snapshot):
    audit_logging.log(
        user, "", operation, user, get_time=fixed_datetime, ip_address="192.168.1.1"
    )

    message = AuditLogEntry.objects.first().message
    assert message["audit_event"]["actor"]["role"] == "OWNER"
    snapshot.assert_match(message)


@pytest.mark.parametrize("operation", list(Operation))
def test_log_user_operation(fixed_datetime, user, user_factory, operation, snapshot):
    other_user = user_factory()

    audit_logging.log(
        user,
        "",
        operation,
        other_user,
        get_time=fixed_datetime,
        ip_address="192.168.1.1",
    )

    message = AuditLogEntry.objects.first().message
    assert message["audit_event"]["actor"]["role"] == "USER"
    snapshot.assert_match(message)


@pytest.mark.parametrize("operation", list(Operation))
def test_log_system_operation(fixed_datetime, user, operation, snapshot):
    audit_logging.log(
        None, "", operation, user, get_time=fixed_datetime, ip_address="192.168.1.1"
    )

    message = AuditLogEntry.objects.first().message
    assert message["audit_event"]["actor"]["role"] == "SYSTEM"
    snapshot.assert_match(message)


@pytest.mark.parametrize("operation", list(Operation))
def test_log_anonymous_role(fixed_datetime, user, operation, snapshot):
    audit_logging.log(
        AnonymousUser(),
        "",
        operation,
        user,
        get_time=fixed_datetime,
        ip_address="192.168.1.1",
    )

    message = AuditLogEntry.objects.first().message
    assert message["audit_event"]["actor"]["role"] == "ANONYMOUS"
    snapshot.assert_match(message)


@pytest.mark.parametrize("status", list(Status))
def test_log_status(fixed_datetime, user, status, snapshot):
    audit_logging.log(
        user,
        "",
        Operation.READ,
        user,
        status,
        get_time=fixed_datetime,
        ip_address="192.168.1.1",
    )

    message = AuditLogEntry.objects.first().message
    assert message["audit_event"]["status"] == status.value
    snapshot.assert_match(message)


def test_log_additional_information(user):
    audit_logging.log(
        user,
        "",
        Operation.UPDATE,
        user,
        additional_information="test",
    )

    message = AuditLogEntry.objects.first().message
    assert message["audit_event"]["additional_information"] == "test"


def test_log_user_with_backend(user, fixed_datetime, snapshot):
    backend_str = "some.auth.Backend"

    audit_logging.log(
        user,
        "some.auth.Backend",
        Operation.READ,
        user,
        get_time=fixed_datetime,
        ip_address="192.168.1.1",
    )

    message = AuditLogEntry.objects.first().message
    assert message["audit_event"]["actor"]["provider"] == backend_str
    snapshot.assert_match(message)
