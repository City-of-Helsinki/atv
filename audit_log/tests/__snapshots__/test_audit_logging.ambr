# serializer version: 1
# name: test_log_actor_uuid
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'USER',
        'user_id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'READ',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': 'be584b90-b256-46f5-83e1-4e6a0f8b4cc3',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_admin_operation[CREATE]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'ADMIN',
        'user_id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'CREATE',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': 'be584b90-b256-46f5-83e1-4e6a0f8b4cc3',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_admin_operation[DELETE]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'ADMIN',
        'user_id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'DELETE',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': 'be584b90-b256-46f5-83e1-4e6a0f8b4cc3',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_admin_operation[READ]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'ADMIN',
        'user_id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'READ',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': 'be584b90-b256-46f5-83e1-4e6a0f8b4cc3',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_admin_operation[UPDATE]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'ADMIN',
        'user_id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'UPDATE',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': 'be584b90-b256-46f5-83e1-4e6a0f8b4cc3',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_anonymous_role[CREATE]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'ANONYMOUS',
        'user_id': '',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'CREATE',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_anonymous_role[DELETE]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'ANONYMOUS',
        'user_id': '',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'DELETE',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_anonymous_role[READ]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'ANONYMOUS',
        'user_id': '',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'READ',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_anonymous_role[UPDATE]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'ANONYMOUS',
        'user_id': '',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'UPDATE',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_owner_operation[CREATE]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'OWNER',
        'user_id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'CREATE',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_owner_operation[DELETE]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'OWNER',
        'user_id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'DELETE',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_owner_operation[READ]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'OWNER',
        'user_id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'READ',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_owner_operation[UPDATE]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'OWNER',
        'user_id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'UPDATE',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_status[FORBIDDEN]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'OWNER',
        'user_id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'READ',
      'origin': 'atv',
      'status': 'FORBIDDEN',
      'target': dict({
        'endpoint': '',
        'id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_status[SUCCESS]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'OWNER',
        'user_id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'READ',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_system_operation[CREATE]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'SYSTEM',
        'user_id': '',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'CREATE',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_system_operation[DELETE]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'SYSTEM',
        'user_id': '',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'DELETE',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_system_operation[READ]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'SYSTEM',
        'user_id': '',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'READ',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_system_operation[UPDATE]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'SYSTEM',
        'user_id': '',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'UPDATE',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_user_operation[CREATE]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'USER',
        'user_id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'CREATE',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': 'be584b90-b256-46f5-83e1-4e6a0f8b4cc3',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_user_operation[DELETE]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'USER',
        'user_id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'DELETE',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': 'be584b90-b256-46f5-83e1-4e6a0f8b4cc3',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_user_operation[READ]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'USER',
        'user_id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'READ',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': 'be584b90-b256-46f5-83e1-4e6a0f8b4cc3',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_user_operation[UPDATE]
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': '',
        'role': 'USER',
        'user_id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'UPDATE',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': 'be584b90-b256-46f5-83e1-4e6a0f8b4cc3',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
# name: test_log_user_with_backend
  dict({
    'audit_event': dict({
      'actor': dict({
        'ip_address': '192.168.1.1',
        'provider': 'some.auth.Backend',
        'role': 'OWNER',
        'user_id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
      }),
      'additional_information': '',
      'date_time': '2020-06-01T00:00:00.000Z',
      'date_time_epoch': 1590969600000,
      'operation': 'READ',
      'origin': 'atv',
      'status': 'SUCCESS',
      'target': dict({
        'endpoint': '',
        'id': '7e564b45-527f-4ea6-92c7-3d39ba05733c',
        'lookup_field': '',
        'type': 'User',
      }),
    }),
  })
# ---
